Feat mike

mjmansfi opened the pull request:
Added BUSCO module and an appropriate test file. The test file will return only a single BUSCO hit, but even this minimal set (a single small yeast chromosome) already takes 4 minutes to complete on my competent desktop. I think an artificial test set should be generated in order for the test to generate more meaningful results more quickly, but for now this will do.

I had Racon files temporarily included in the commit history so that I could pull the Racon files, however development of Racon should properly be moved to a different branch and these have been removed.

chris-cheshire commented on tools/busco/Dockerfile:
```diff
@@ -0,0 +1,3 @@
+FROM ezlabgva/busco:v4.1.3_cv1
```
why not just reference the official image?

chris-cheshire commented on tools/busco/main.nf:
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env nextflow
+
+// Specify DSL2
+nextflow.preview.dsl = 2
```
This is old syntax `nextflow.enable.dsl=2` use this where you can (we still need to go through and change this)

chris-cheshire replied:
dont worry ive done this in a PR for all

chris-cheshire commented on tools/busco/main.nf:
```diff
@@ -0,0 +1,43 @@
+                      if (opts.publish_results == "none") null
+                      else filename }
+
+    container "ezlabgva/busco:v4.1.3_cv1"
```
ahh i see you have! just delete the docker file then

chris-cheshire commented on tools/busco/main.nf:
```diff
@@ -0,0 +1,43 @@
+        tuple val(meta), path(genome)
+
+    output:
+        tuple val(meta), path("busco"), emit: busco_genome_out
```
we had a variable name meeting (I know), can you change the name to describe the files emitted? our preferred names are things like `sam` and `bam` etc.

mjmansfi replied:
I would like some clarification for this comment. Is the goal to make the `emit` names descriptive (describing the files, as you say), or is the goal to make them compact and simple? 

A simple `emit` name is fine if the process name is specific since the distinction between related outputs will be clear (e.g., `initial_alignment.out.bam` vs. `highly_modified_and_subsetted_alignment.out.bam`). But if both the process name and the emit name are generic and short, it will be harder to distinguish outputs that happen to share an output type (e.g., `align.out.bam` vs. `align.out.bam`).

In this case the output of the process is a directory containing a bunch of files and subdirectories. I could split every expected subdirectory and file into separate `emit` statements, but there are many, and none of outputs will be specifically used as input to later processes. The purpose of this module is essentially QC; while none of its outputs are particularly useful, it is necessary to generate them and they are worth keeping. 

Having described this, is it better for the emit name to be simply _busco_, or something like _busco_genome_output_directory_?

chris-cheshire replied:
well based on what you are saying, cant you just call it `qc`? Or, if its a reporting output that is not consumed by a downstream process, we usually call this `report`

chris-cheshire commented on tools/busco/test/main.nf:
```diff
@@ -0,0 +1,41 @@
+
+workflow {
+    busco_genome( params.modules['busco_genome'], ch_fasta )
+    busco_genome.out.busco_genome_out | view
```
can you add a channel count assertion?

mjmansfi commented:
I have changed `emit`s to just `report` to simplify things as suggested. I also changed the test set to the set included in BUSCO by default. This test set yields a meaningful BUSCO result and has a small file size - but it takes 8 minutes to run on my work setup. 

Increasing the number of CPUs in the `nextflow.config` for the test module will speed things up, but perhaps this is undesirable for other reasons.

chris-cheshire commented:
Yeah we only get one core on github actions - any easy way to reduce test dataset size?

chris-cheshire commented:
can you also add busco to the module test file as well?

mjmansfi commented:
I'm going to close this. In the future I will use the `feat-mike` branch only as a sandbox, and open new branches for different modules as they are required.
